### PR TITLE
UHF-4531: Address search changes

### PR DESCRIPTION
**Pull request by tapsal**
# Address search changes [UHF-4531](https://helsinkisolutionoffice.atlassian.net/browse/UHF-4531)
Fixes some address search sorting issues and adds some requested changes.

## What was done
* Fixed sort order when showing entities and/or using pager.
* Shows the distance in kilometers. Shown minimum distance is 0,1 km.
* Added views area plugin for showing search information.

## How to install
* Make sure your instance is up and running on latest dev branch.
  * It might be easiest to test with the Kasko instance.
* Update the TPR module:
    * `composer require drupal/helfi_tpr:dev-UHF-4531-address-search-changes`
* Go to `/admin/modules` and enable "HELfi Address search".
* Edit some view that lists units, e.g. the new Daycare view at Kasko (see: [separate PR](https://github.com/City-of-Helsinki/drupal-helfi-kasvatus-koulutus/pull/116)).
* Make sure the view has `Computed distance` field with default settings.
* Make sure the view has `Address search` filter with default settings.
* Check which template is used when showing the individual view results. For example, with the Kasko's Daycare view, the template file is `hdbt_subtheme/templates/module/helfi_tpr/tpr-unit.html.twig`.
  * Make sure the template has `{{ content.distance }}` for showing the distance.
* Clear caches: `drush cr`
* Make sure you have some page, which includes a paragraph for the unit search (e.g. Daycare search) and it shows some units.

## How to test
* Edit the view by adding a "Address search information" area to the header.
  * Add some informative text for successful and failed searches. 
* Go to the page that contains the view.
* Search with different addresses.
* Check that successful search shows the informative text for successful searches.
* Check that failed search (address not found) shows the informative text for failed searches.
* Check that the distance is shown as kilometers. Check that the smallest distance is 0,1 km.
* Check that results are sorted correctly: the nearest units are shown first.
* Check that the sorting works in different conditions:
  * When showing all results at once.
  * When using different pagers (e.g. pages and infinite scroll).
  * When cache is enabled.

### CODE REVIEW

File: modules/helfi_address_search/helfi_address_search.module [hunk 1]
```diff
@@ -33,14 +34,36 @@ function helfi_address_search_views_data_alter(array &$data) {
...
+    'area' => [
+      'id' => 'address_search_info',
+    ],
+  ];
+}
+
+/**
+ * Implements hook_views_pre_execute().
+ */
+function helfi_address_search_views_pre_execute(ViewExecutable $view) {
```
Comment (tuutti) [inline modules/helfi_address_search/helfi_address_search.module:50]: Missing return type (`void`).

File: modules/helfi_address_search/helfi_address_search.module [hunk 2]
```diff
@@ -33,14 +34,36 @@ function helfi_address_search_views_data_alter(array &$data) {
...
+  // Query all results as the sorting and paging is done with
+  // AddressSearch::getSortedResultsByAddress().
+  $view->query->setLimit(NULL);
+  $view->query->setOffset(NULL);
 }
 
 /**
  * Implements hook_views_pre_render().
  */
 function helfi_address_search_views_pre_render(ViewExecutable $view) {
```
Comment (tuutti) [inline modules/helfi_address_search/helfi_address_search.module:64]: Missing return type (`void`).

File: modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ *     "float"
+ *   }
+ * )
+ */
+class KiloFormatter extends DecimalFormatter {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function defaultSettings() {
```
Comment (tuutti) [inline modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php:30]: Missing return type.

File: modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      'scale' => 1,
+      'prefix_suffix' => TRUE,
+      'minimum_value' => 0.1,
+    ] + parent::defaultSettings();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(array $form, FormStateInterface $form_state) {
```
Comment (tuutti) [inline modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php:43]: Missing return type.

File: modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      '#weight' => 7,
+    ];
+
+    return $elements;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function numberFormat($number) {
```
Comment (tuutti) [inline modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php:63]: Missing return type and `$number` is missing argument type.

File: modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    if (isset($this->view->element['address_search_succeed'])) {
+      return $this->view->element['address_search_succeed'];
+    }
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function defineOptions() {
```
Comment (tuutti) [inline modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php:37]: Missing return type.

File: modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        'format' => ['default' => NULL],
+      ],
+    ];
+    return $options;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildOptionsForm(&$form, FormStateInterface $form_state) {
```
Comment (tuutti) [inline modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php:57]: Missing return type (`void`) and `$form` is missing argument type

File: modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+      '#rows' => 6,
+      '#format' => $this->options['failed']['format'] ?? filter_default_format(),
+      '#editor' => FALSE,
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function render($empty = FALSE) {
```
Comment (tuutti) [inline modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php:82]: Missing return type. `$empty` missing argument type.

**Review by tuutti — CHANGES_REQUESTED** [review]

File: modules/helfi_address_search/helfi_address_search.module (hunk 1 above)
Comment (tapsal) [reply modules/helfi_address_search/helfi_address_search.module]: Fixed.

File: modules/helfi_address_search/helfi_address_search.module (hunk 2 above)
Comment (tapsal) [reply modules/helfi_address_search/helfi_address_search.module]: Fixed.

File: modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php (hunk 3 above)
Comment (tapsal) [reply modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php]: Fixed.

File: modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php (hunk 4 above)
Comment (tapsal) [reply modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php]: Fixed.

File: modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php (hunk 5 above)
Comment (tapsal) [reply modules/helfi_address_search/src/Plugin/Field/FieldFormatter/KiloFormatter.php]: Fixed.

File: modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php (hunk 6 above)
Comment (tapsal) [reply modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php]: Fixed.

File: modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php (hunk 7 above)
Comment (tapsal) [reply modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php]: Fixed.

File: modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php (hunk 8 above)
Comment (tapsal) [reply modules/helfi_address_search/src/Plugin/views/area/AddressSearchInfo.php]: Fixed.

**Review by tuutti — APPROVED** [review]

**Review by teroelonen — APPROVED** [review]
This is gooooood to go 🦖

**Pull request merged**